src/Build/Evaluation/LazyItemEvaluator.RemoveOperation.cs

```diff
@@ -55,13 +56,23 @@ protected override ImmutableList<I> SelectItems(ImmutableList<ItemData>.Builder
                 var items = ImmutableHashSet.CreateBuilder<I>();
                 foreach (ItemData item in listBuilder)
                 {
-                    if (_matchOnMetadata.IsEmpty ? _itemSpec.MatchesItem(item.Item) : _itemSpec.MatchesItemOnMetadata(item.Item, _matchOnMetadata, _matchOnMetadataOptions))
+                    if (_matchOnMetadata.IsEmpty ? _itemSpec.MatchesItem(item.Item) : MatchesItemOnMetadata(item.Item))
                         items.Add(item.Item);
                 }
 
                 return items.ToImmutableList();
             }
 
+            private bool MatchesItemOnMetadata(I item)
+            {
+                if (metadataSet == null)
```

Comment:
I'd keep MetadataSet as a hidden implementation detail in ItemSpec. The ItemSpec users do not need to know about it.

Reply:
They don't have to know about it anyway. The methods that call it in ItemGroupIntrinsicTask and LazyItemEvaluator.RemoveOperation are both private. I count that as close enough.

Reply:
By users I meant Itemspec API users. When somebody tries to understand the code in ItemGroupIntrinsicTask  or in LazyItemEvaluator.RemoveOperation they'll get hit with two concepts they have to understand: Itemspec and MetadataSet. If you hide MetadataSet inside Itemspec it becomes an implementation detail that people don't have to care about unless they're interested in how Itemspec works.